### DISCUSSION TITLE
How to create unit tests ?

### DISCUSSION THREAD

**Discussion by dufguix** [serverpod/serverpod#1958] [Q&A] [answered]
I mean how to handle the context (aka session object) during a unit test ?

**Comment by Isakdl** [comment, marked as answer]
I would avoid dealing with the session object in a unit test as it will require I/O calls that would imply heavy mocking. 

However, if you need to do integration tests where you do want to use the session object you can create your own session object from a serverpod instance.


Example:

```dart
test('description', () async {
    final args = <String>[];
    final serializationManager = Protocol();
    final endpoints = Endpoints();
    final serverpod = Serverpod(
      args,
      serializationManager,
      endpoints,
    );
    await serverpod.start();
    final session = await serverpod.createSession();

    // Test code goes here.

    await session.close();
    await serverpod.shutdown(exitProcess: false);
  });
```

`Endpoints` and `Protocol` should be imported from the generated code in the server. 

Keep in mind that you will need to have a database running.

**Reply by dufguix** [reply]
crystal clear. thanks.

Is there a serverpod instance without the need of IO access with in-ram DB, in-ram global cache, in-ram S3 storage ?

**Reply by Isakdl** [reply]
Not yet, would love to add this functionality in the future though.

**Reply by dufguix** [reply]
Okay, I'll search on docker side. 
With package like [Docker2](https://pub.dev/packages/docker2) to create and remove by test.
